kernel/kernel-api/src/main/java/io/delta/kernel/internal/rowtracking/RowTracking.java

```diff
@@ -0,0 +1,160 @@
+
+                  AddFile addFile = AddFile.fromRow(row.getStruct(ADD_FILE_ORDINAL));
+
+                  // Assign base row ID if missing
+                  if (!addFile.getBaseRowId().isPresent()) {
+                    final long numRecords = getNumRecords(addFile);
+                    addFile = addFile.withNewBaseRowId(currRowIdHighWatermark[0] + 1L);
+                    currRowIdHighWatermark[0] += numRecords;
+                  }
+
+                  // Assign default row commit version if missing
+                  if (!addFile.getDefaultRowCommitVersion().isPresent()) {
+                    addFile = addFile.withNewDefaultRowCommitVersion(commitVersion);
+                  }
+
+                  // Return a new AddFile row with assigned baseRowId/defaultRowCommitVersion
+                  return SingleAction.createAddFileSingleAction(addFile.toRow());
+                });
+      }
+    };
+  }
+
+  /**
+   * Emits a {@link DomainMetadata} action if the row ID high watermark has changed due to newly
+   * processed {@link AddFile} actions.
+   *
+   * @param protocol the protocol to check for row tracking support
+   * @param snapshot the current snapshot of the table
+   * @param dataActions the iterable of data actions that may update the high watermark
+   */
+  public static Optional<DomainMetadata> createNewHighWaterMarkIfNeeded(
+      Protocol protocol, SnapshotImpl snapshot, CloseableIterable<Row> dataActions) {
+    if (!TableFeatures.isRowTrackingSupported(protocol)) {
+      return Optional.empty();
+    }
+
+    final long prevRowIdHighWatermark = readRowIdHighWaterMark(snapshot);
+    // Use a one-element array here to allow for mutation within the lambda.
+    final long[] newRowIdHighWatermark = {prevRowIdHighWatermark};
+
+    dataActions.forEach(
+        row -> {
+          if (!row.isNullAt(ADD_FILE_ORDINAL)) {
+            newRowIdHighWatermark[0] +=
+                getNumRecords(AddFile.fromRow(row.getStruct(ADD_FILE_ORDINAL)));
+          }
+        });
+
+    return (newRowIdHighWatermark[0] != prevRowIdHighWatermark)
+        ? Optional.of(new RowTrackingMetadataDomain(newRowIdHighWatermark[0]).toDomainMetadata())
+        : Optional.empty();
+  }
+
+  /**
+   * Reads the current row ID high watermark from the snapshot, or returns a default value if
+   * missing.
+   *
+   * @param snapshot the snapshot to read
+   * @return the row ID high watermark
+   */
+  public static long readRowIdHighWaterMark(SnapshotImpl snapshot) {
+    return RowTrackingMetadataDomain.fromSnapshot(snapshot)
+        .map(RowTrackingMetadataDomain::getRowIdHighWaterMark)
+        .orElse(RowTrackingMetadataDomain.MISSING_ROW_ID_HIGH_WATERMARK);
+  }
+
+  /**
+   * Retrieves the number of records from the AddFile's statistics. We error out if statistics are
+   * missing.
+   *
+   * @param addFile the AddFile action
+   * @return the number of records
+   */
+  private static long getNumRecords(AddFile addFile) {
```

Comment:
My idea was to make it also throw `rowIDAssignmentWithoutStats` in case of missing stats to simply some row tracking code, which is very specific to row tracking and would be inappropriate for a getter method of AddFile itself.

Now I've put the getter `public Optional<Long> getNumRecords()` to the AddFile itself, and have a helper method in `RowTracking.java`:
```
private static long getNumRecordsOrThrow(AddFile addFile) {
    return addFile.getNumRecords().orElseThrow(DeltaErrors::rowIDAssignmentWithoutStats);
}
```
Does this look good to you?